druid-shell/examples/invalidate.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+
+use std::time::Instant;
+
+use druid_shell::kurbo::{Point, Rect};
+use druid_shell::piet::{Color, Piet, RenderContext};
+
+use druid_shell::{Application, WinHandler, WindowBuilder, WindowHandle};
+
+struct InvalidateTest {
+    handle: WindowHandle,
+    size: (f64, f64),
+    start_time: Instant,
+    color: Color,
+    rect: Rect,
+}
+
+impl InvalidateTest {
+    fn update_color_and_rect(&mut self) {
+        let time_since_start = (Instant::now() - self.start_time).as_nanos();
+        let (r, g, b, _) = self.color.as_rgba_u8();
+        self.color = match (time_since_start % 2, time_since_start % 3) {
+            (0, _) => Color::rgb8(r.wrapping_add(10), g, b),
+            (_, 0) => Color::rgb8(r, g.wrapping_add(10), b),
+            (_, _) => Color::rgb8(r, g, b.wrapping_add(10)),
+        };
+
+        self.rect.x0 = (self.rect.x0 + 5.0) % self.size.0;
+        self.rect.x1 = (self.rect.x1 + 5.5) % self.size.0;
+        self.rect.y0 = (self.rect.y0 + 3.0) % self.size.1;
+        self.rect.y1 = (self.rect.y1 + 3.5) % self.size.1;
+    }
+}
+
+impl WinHandler for InvalidateTest {
+    fn connect(&mut self, handle: &WindowHandle) {
+        self.handle = handle.clone();
+    }
+
+    fn paint(&mut self, piet: &mut Piet, rect: Rect) -> bool {
+        self.update_color_and_rect();
+        piet.fill(rect, &self.color);
+
+        self.handle.invalidate_rect(self.rect);
+        false
+    }
```

Comment:
on mac, invalidation calls that happen during drawing are ignored.

I would suggest changing this to use a timer, something like:
```suggestion
    fn connect(&mut self, handle: &WindowHandle) {
        self.handle = handle.clone();
        self.handle.request_timer(std::time::Duration::from_millis(60));
    }
    
    fn timer(&mut self, _id: TimerToken) {
        self.update_color_and_rect();
        self.handle.invalidate_rect(self.rect);
        self.handle.request_timer(std::time::Duration::from_millis(60));
    }

    fn paint(&mut self, piet: &mut Piet, rect: Rect) -> bool {
        piet.fill(rect, &self.color);
        false
    }
```